ASTGen: An assortment of strides and improvements

AnthonyLatsis opened the pull request:
This pull request focuses on declaration translation, extrapolating preparatory work to the rest of the module for consistency and testing.

AnthonyLatsis committed:
ASTGen: Translate associated type declarations

Plus tweak `DefaultDefinitionTypeRequest` caching to support querying the
cached type when dumping. This fixes a crash where type computation is
triggered in the dumper before import resolution in `-dump-parse` mode.

AnthonyLatsis committed:
ASTGen: Translate enumeration declarations

AnthonyLatsis committed:
ASTGen: Translate enum case declarations

AnthonyLatsis committed:
ASTGen: Translate enum element payloads

AnthonyLatsis committed:
ASTGen: Set brace locations for iterable declaration contexts

AnthonyLatsis committed:
ASTGen: Fix enough source locations that we can start diffing AST dumps

... and tidy up adjacent code along the way.

AnthonyLatsis committed:
ASTGen: Translate class generic parameter clauses

AnthonyLatsis committed:
ASTGen: Translate actor declarations

AnthonyLatsis committed:
ASTGen: Translate protocol declarations

AnthonyLatsis committed:
ASTGen: Translate inheritance clauses

AnthonyLatsis committed:
ASTGen: Translate extension declarations

AnthonyLatsis committed:
ASTGen: Translate function generic parameter clauses

AnthonyLatsis committed:
ASTGen: Cache parsed `IterableDeclContext` members using the request evaluator

...instead of adding them to the iterable context lest they be re-parsed using
the legacy parser and later re-added through `IterableDeclContext::loadAllMembers`.

AnthonyLatsis committed:
[NFC] ASTGen: Don't unnecessarily bridge `TypeAliasDecl::setUnderlyingTypeRepr`

AnthonyLatsis committed:
[NFC] ASTGen: Extract `TokenSyntax` → `BridgedIdentifier` mapping into a method

AnthonyLatsis committed:
ASTGen: Translate operator declarations

AnthonyLatsis committed:
ASTGen: Translate generic where clauses

AnthonyLatsis committed:
ASTGen: Translate deinitializer declarations

AnthonyLatsis committed:
[NFC] ASTGen: Extract `ParamDecl` and `ParameterList` translation into a new file

AnthonyLatsis committed:
ASTGen: Translate initializer declarations

AnthonyLatsis committed:
ASTGen: Translate import declarations

AnthonyLatsis committed:
ASTGen: Skip leading trivia when forming bridged source locations

AnthonyLatsis committed:
ASTGen: Translate default arguments

AnthonyLatsis committed:
ASTGen: Translate precedence group declarations

AnthonyLatsis committed:
[NFC] ASTGen: Optimize and flatten the transform + `withBridgedArrayRef` pattern via bump allocation

To encourage usage of  lazy transformations, define `bridgedArrayRef`
on `LazyCollectionProtocol`.

AnthonyLatsis committed:
[NFC] ASTGen: Remove unused `SwiftParser` imports

AnthonyLatsis commented on test/ASTGen/verify-parse.swift:
```diff
@@ -1,33 +1,74 @@
-// RUN: %target-run-simple-swift(-enable-experimental-feature SwiftParser -enable-experimental-feature ParserASTGen)
-// RUN: %target-run-simple-swift(-enable-experimental-feature ASTGenTypes)
+
+// RUN: %empty-directory(%t)
+// RUN: %target-swift-frontend %s -dump-parse -disable-availability-checking -enable-experimental-feature ParserASTGen > %t/astgen.ast
+// RUN: %target-swift-frontend %s -dump-parse -disable-availability-checking > %t/cpp-parser.ast
+// RUN: %diff -u %t/astgen.ast %t/cpp-parser.ast
```
This is of course ersatz — we ought to implement an AST matcher.

DougGregor replied:
We should, but it's a big task and the approach you're using here can provide real value along the way.

slavapestov commented:
Nice!

AnthonyLatsis commented:
@swift-ci please smoke test

AnthonyLatsis commented:
@swift-ci please smoke test macOS

AnthonyLatsis commented:
@swift-ci please smoke test macOS

bnbarham reviewed (commented):
Thanks @AnthonyLatsis. Big fan of your operator test choices :)

DougGregor commented on lib/AST/ASTDumper.cpp:
```diff
@@ -656,9 +656,14 @@ namespace {
+      } else if (auto repr = decl->getDefaultDefinitionTypeRepr()) {
+        repr->print(OS);
+      } else {
+        OS << "<not computed>";
```
"Not computed" and "nonexistent" are different things here. Can we differentiate in the AST dumper output?

DougGregor commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -1,162 +1,503 @@
-    var nameText = node.identifier.text
-    let name = nameText.withBridgedString { bridgedName in
-      return ASTContext_getIdentifier(ctx, bridgedName)
+    let (name, nameLoc) = self.bridgedIdentifierAndSourceLoc(for: node.identifier)
```
This is a whole lot nicer

DougGregor commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -1,162 +1,503 @@
+    case .keyword(.none): self = .none
+    case .keyword(.left): self = .left
+    case .keyword(.right): self = .right
+    default: fatalError("Unknown precedence group associativity value")
```
Should this produce a diagnostic rather than failing? The incoming syntax tree could be ill-formed.

AnthonyLatsis replied:
Do we not expect a `SwiftParserDiagnostics` pass to catch this sort of thing? I.e. why do we want to allow for translation of ill-formed syntax trees?

bnbarham replied:
ASTGen should handle any swift-syntax tree, whether from the parser or not. There's a few cases that currently have `!` that really should not as well. @ahoppen as I know he has strong opinions on this :)

AnthonyLatsis replied:
Oh I get it know, this is about ill-formed trees, not source syntax. So what we want is a generic error about the syntax tree being invalid, I suppose? We don’t seem to currently have a way to propagate out an error either.

AnthonyLatsis replied:
> ASTGen should handle any swift-syntax tree, whether from the parser or not. There's a few cases that currently have ! that really should not as well.

Does SwiftSyntax have a way to verify nodes? We could have an interceptor or verifier pre-pass rather than having to recover from every corner where the model is too permissive in the middle of translation.

ahoppen replied:
> is there any value in attempting to recover from an invalid syntax tree?

There absolutely is! For example if you write `var x Int`, the parser will recover and introduce the `:` as a missing token and we should continue type checking based on that assumption to emit any diagnostics type checking might produce.

> It’s almost as if this is starting to resemble the mess that the legacy parser is

Yes, it’s unfortunate but these checks need to live somewhere 😬 

> Is there any reason we do not want a separate validation pass preceding translation (e.g. `sourceFile.validate()`), using the validation mechanics that are built in to SwiftSyntax?

If we split validation and the translation into the C++ AST into two separate phases, the two phases will inevitably diverge and we’ll end up in situations where ASTGen makes assumptions about things that were should have been detected in `validate()`, which introduces some kind of statefulness to the phase of the compiler at which you are viewing the SwiftSyntax tree. And IMO that’s a major problem with the C++ tree at the moment: Type checking / SILGen / … all make assumptions about which shape the tree might be in at that point and these assumptions to get broken on a regular basis, usually resulting in assertion failures / crashes.

AnthonyLatsis replied:
> There absolutely is! For example if you write `var x Int`, the parser will recover and introduce the : as a missing token and we should continue type checking based on that assumption to emit any diagnostics type checking might produce.

I was thinking of non-parsed trees fed by a fuzzer or something, but I forgot you mentioned macro expansions as a possible input too. It makes perfect sense to recover when parsing written code though.

> If we split validation and the translation into the C++ AST into two separate phases, the two phases will inevitably diverge and we’ll end up in situations where ASTGen makes assumptions about things that were should have been detected in `validate()`.

My point is more about reusing existing validation logic. Even starting each `visit` method with a SwiftSyntax validation and error handling the result would be so much cleaner than reinventing it the way we do here. It feels sensible to me for ASTGen to trust SwiftSyntax and assume that it is being handed a valid structure per the language grammar. Personally, I think this assumption is less risky in the long run than the assumption that the SwiftSyntax and ASTGen validations are consistent.

ahoppen replied:
> My point is more about reusing existing validation logic. Even starting each `visit` method with a SwiftSyntax validation and error handling the result would be so much cleaner than reinventing it the way we do here. It feels sensible to me for ASTGen to trust SwiftSyntax and assume that it is being handed a valid structure per the language grammar. Personally, I think this assumption is less risky in the long run than the assumption that the SwiftSyntax and ASTGen validations are consistent.

Whenever we make assumptions on the syntax tree, I think those should be implemented in the type system and in many cases, I believe we succeed in that. Operator declaration parsing is a really weird one because of all the custom grammar that it has.

AnthonyLatsis replied:
> Whenever we make assumptions on the syntax tree, I think those should be implemented in the type system

Definitely!

> Operator declaration parsing is a really weird one because of all the custom grammar that it has

Do you mean precedence group parsing? Did you consider replacing `groupAttributes` with an optional child for every of the 4 kinds of attributes, and thus encoding that they must appear at most once?

ahoppen replied:
> Do you mean precedence group parsing? 

Oh, yes, sorry, I mixed them up.

> Did you consider replacing `groupAttributes` with an optional child for every of the 4 kinds of attributes, and thus encoding that they must appear at most once?

Not so far, but I think that would make sense.

AnthonyLatsis replied:
Ah, right. How ugly would it be to not actually enforce their order in the layout? Say, the initializer always inserts the attributes in a fixed order, but one could pass `nil`s in the initializer and then insert the attributes in the appropriate order via the setters if said order is important. The getters and setters would then work with a specific subrange of the layout of size 4 which is nilled out by default.

ahoppen replied:
I’m not entirely sure what you mean. Do you have a short example of how you think the API should look like?

AnthonyLatsis replied:
An optional child for each attribute kind as originally suggested, with accessors similar to these: 
```swift
  public var higherThanRelation: PrecedenceGroupRelationSyntax? {
      get {
        for index in 11...14 {
          guard let childData = data.child(at: index) else {
            // No data for this attribute (attributes are stored contiguously, i.e., no 'nil' data in between attributes).
            break
          }

          if let value = PrecedenceGroupRelationSyntax(Syntax(childData)) {
            return value
          }
        }

        return nil
      }

      set(value) {
        for index in 11...14 {
          guard let rawChild = data.rawChild(at: index) else {
            // No data for this attribute (attributes are stored contiguously, i.e., no 'nil' data in between attributes).
            // Stick it in at this index.
            self = PrecedenceGroupDeclSyntax(data.replacingChild(at: index, with: value?.data, arena: SyntaxArena()))
            return
          }

          if rawChild.kind == .precedenceGroupRelation {
            if let value {
              // Simply replace the data at this index.
              self = PrecedenceGroupDeclSyntax(data.replacingChild(at: index, with: value.data, arena: SyntaxArena()))
              return
            } else {
              self = PrecedenceGroupDeclSyntax(data.replacingChild(at: index, with: nil, andShiftingItTo: 14, arena: SyntaxArena()))
            }
          }
        }
      }
    }
  }
```

Another possibility is to statically feature the 4 optional child attributes in `PrecedenceGroupAttributeListSyntax`, as they would have been in `PredenceGroupDeclSyntax`, perhaps retaining the conformance to `SyntaxCollection`. This way no changes to `PredenceGroupDeclSyntax` are necessary.

ahoppen replied:
> An optional child for each attribute kind as originally suggested, with accessors similar to these.

The issue I see with this design is that a client of SwiftSyntax might create a `PrecedenceGroupDeclSyntax` that has two `higherThan` relations in the `groupAttributes` and the `higherThanRelation` getter would only return the first.

Similarly, if you set a `higherThan` relation using this setter, you can’t control where it is being inserted if no `higherThan` relation exists yet. So

```swift
let pg = PrecedenceGroupDeclSyntax()
pg.higherThanRelation = // …
pg.assignment = // ...
```

would give you a different node than

```swift
let pg = PrecedenceGroupDeclSyntax()
pg.assignment = // ...
pg.higherThanRelation = // …
```

and that’s quite surprising IMO.

DougGregor commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -1,162 +1,503 @@
+  func visit(_ node: PrecedenceGroupDeclSyntax) -> ASTNode {
+    let (name, nameLoc) = self.bridgedIdentifierAndSourceLoc(for: node.identifier)
+
+    struct PrecedenceGroupBody {
```
For precedence groups and operators, I wonder if we should instead build on top of the `SwiftOperators` library, which has a mapping from the syntax tree into a more semantic representation (`Operator` and `PrecedenceGroup`).

AnthonyLatsis replied:
Apparently the inits from `SwiftOperators` assume the nodes are valid, so we would have to do most of the same work in `ASTGen` to prevalidate the nodes and diagnose issues before calling these inits, like switching on the fixity of an operator or iterating over the attributes of a precedence group. Do you believe there is something to gain from building on top of `SwiftOperators` regardless?

DougGregor commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -1,162 +1,503 @@
+    case .keyword(.protocol): self = .protocol
+    case .keyword(.var), .keyword(.let): self = .var
+    case .keyword(.func): self = .func
+    default: fatalError("Unknown import kind")
```
It seems like this could trip up on invalid syntax trees.

DougGregor commented on lib/ASTGen/Sources/ASTGen/Exprs.swift:
```diff
@@ -1,17 +1,22 @@
+          self.bridgedSourceLoc(for: node.rightBrace)
+        )
+
+        return ClosureExpr_create(self.ctx, body, self.declContext)
```
Can you put a TODO or FIXME in here to translate the signature of a closure? (parameter list, captures, etc.)

DougGregor commented on lib/ASTGen/Sources/ASTGen/Literals.swift:
```diff
@@ -1,47 +1,52 @@
 extension ASTGenVisitor {
   public func visit(_ node: StringLiteralExprSyntax) -> ASTNode {
-    let loc = bridgedSourceLoc(for: node)
+    let openDelimiterOrQuoteLoc = self.bridgedSourceLoc(for: node.openDelimiter ?? node.openQuote)
```
I know it was like this before, but I'd appreciate a TODO/FIXME to handle string interpolations.

DougGregor reviewed (commented):
This is really quite awesome! A few minor comments here and there, but I love that you've increased the coverage of ASTGen so far.

It would be fun to follow this up with the infrastructure to redirect the C++ parser over to ASTGen  for specific declaration kinds (e.g., import decls, operators, other small stuff), like I did in https://github.com/apple/swift/pull/66033, which lets you run a whole bunch of code through a narrow slice of ASTGen.

DougGregor commented on lib/AST/Decl.cpp:
```diff
@@ -6022,7 +6022,8 @@ ProtocolDecl::ProtocolDecl(DeclContext *DC, SourceLoc ProtocolLoc,
     : NominalTypeDecl(DeclKind::Protocol, DC, Name, NameLoc, Inherited,
                       nullptr),
       ProtocolLoc(ProtocolLoc),
-      PrimaryAssociatedTypeNames(PrimaryAssociatedTypeNames) {
```
I think @bnbarham is right that we should keep it on the bridging side, if only to keep big ASTGen PRs like this one from causing changes elsewhere in the compiler.

AnthonyLatsis replied:
I did this because it felt counterintuitive and fragile to rely on callers to pre-copy, and I forgot to myself. Worst thing we copy twice; I think that’s a better state of affairs than UB, but I won’t insist.

> and we're doing that for inheritedTypes now as well.

I dropped the ball on inherited types because there are a lot more places to adjust in this case and that seemed too much of a detour for this PR. I would be happy to revisit in a follow-up though.

bnbarham replied:
I'd say that's a much larger discussion, though probably one we should have. For now as @DougGregor said I'd prefer we we kept large ASTGen implementations like this PR as self contained as they can be.

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/Misc.swift:
```diff
@@ -75,3 +61,39 @@ extension String {
   }
 }
+
+extension ASTGenVisitor {
```
Sounds good. Do you mind keeping `Misc.swift` for misc visits?

bnbarham replied:
If you strongly prefer, personally I think `ASTGen.swift` is perfectly reasonable for them. Ideally we wouldn't have a `Misc.swift` since it ends up being a dumping ground 😅

AnthonyLatsis commented on include/swift/AST/CASTBridging.h:
```diff
@@ -226,6 +226,11 @@ typedef struct BridgedTypeAttributes {
   void *raw;
 } BridgedTypeAttributes;
 
+struct BridgedIdentifierAndSourceLoc {
+  BridgedIdentifier name;
+  BridgedSourceLoc nameLoc;
+};
```
Ah you mean in the bridging constructors? It actually never occurred to me, but sounds like a great idea! (for another pull request though; this one is getting too meaty).

bnbarham replied:
Yeah, there's lots that take a name + loc that could now take this instead (which would also simplify the swift side). I'm perfectly okay with that being in another PR.

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/Generics.swift:
```diff
@@ -39,51 +39,37 @@ extension ASTGenVisitor {
     _ node: GenericParameterClauseSyntax,
     action: (BridgedArrayRef, BridgedArrayRef) -> T
   ) -> T {
-    var params = [UnsafeMutableRawPointer]()
```
Not a fan of this method in general, I got rid of it down the line in the "translate `where` clauses" commit . That said, I agree the functional version is not as straightforward, but.. that bad? Could the cluttered diff have affected your impression?

> Do you have any numbers on what the performance gain here is?

The motivation for `map(_: withBridgedArrayRef:)` is as much fast path provision as encapsulation, but regardless, I’m a kind of person that doesn’t need a bad metric score to do a refactor of this sort and scope. This here is really just me being consistent across all occurrences of the `map` + `withBridgedArrayRef` pattern though.

bnbarham replied:
I just commented on the first one I saw here, but my comment was more generally about `map(_: withBridgedArrayRef:)` than this method itself. Personally I find this much harder to read because of the nesting, but if I'm the only one that thinks that... I don't have a better alternative 🤷

ahoppen commented:
Just a quick comment: I’d like to review the PR but won’t get to it until Monday/Tuesday. I’d appreciate if you could leave it open until then.

AnthonyLatsis commented:
> I’d like to review the PR but won’t get to it until Monday/Tuesday. I’d appreciate if you could leave it open until then.

No problem! I would hate to land this without a review from you.

ahoppen commented on lib/ASTGen/Sources/ASTGen/ParameterClause.swift:
```diff
@@ -0,0 +1,111 @@
+
+    let firstName: BridgedIdentifier
+    if let name = node._firstName {
+      if name.text == "_" {
```
Comparing the `tokenKind` is much cleaner than going to the text-representation.

```suggestion
      if name.tokenKind == .wildcard {
```

ahoppen commented on lib/ASTGen/Sources/ASTGen/ParameterClause.swift:
```diff
@@ -0,0 +1,111 @@
+// MARK: - ParamDecl
+
+fileprivate protocol ValueParameterSyntax: SyntaxProtocol {
+  var _firstName: TokenSyntax? { get }
```
I’m not a fan of these underscored properties. Underscored properties to me are always things that aught to be private but aren’t for some stupid reason. 

I would prefer

```swift
/// The `firstName` with optional type. 
///
/// This is the lowest denominator between `FunctionParameterSyntax` and `EnumCaseParameterSyntax`.
var optionalFirstName: TokenSyntax? { get }
```

AnthonyLatsis replied:
These requirements are a sort of hack around our lack of support for covariant witnesses. I used underscores to highlight this, but I am happy to rename and add a FIXME instead.

ahoppen commented on lib/ASTGen/Sources/ASTGen/Types.swift:
```diff
@@ -244,38 +237,25 @@ extension ASTGenVisitor {
-    return elements.withBridgedArrayRef { elements in
+
+      return BridgedTupleTypeElement(
+        Name: name,
+        NameLoc: nameLoc,
+        SecondName: secondName,
+        SecondNameLoc: secondNameLoc,
+        UnderscoreLoc: nil, /*N.B. Only important for SIL*/
+        ColonLoc: self.bridgedSourceLoc(for: element.colon),
+        Type: type,
+        TrailingCommaLoc: self.bridgedSourceLoc(for: element.trailingComma)
```
Just me complaining, no action necessary for you in this PR: I know that you didn’t introduce `BridgedTupleTypeElement` but the members of it should really be lowercase so I don’t need to look at uppercase parameter labels in Swift.

ahoppen reviewed (commented):
Here’s my first round of review. You’re getting a bit of my complaints with the current ASTGen design but I think that I managed to keep my comments somewhat focused so that I don’t require you to do a massive refactoring first.

rintaro commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -1,162 +1,503 @@
+    let (name, nameLoc) = self.bridgedIdentifierAndSourceLoc(for: node.identifier)
 
-    return .decl(out.decl)
+    let declAndDeclContext = (node.inheritanceClause?.inheritedTypeCollection).map {
+      self.visit($0.typeName).rawValue
+    } withBridgedArrayRef: { inheritedTypes in
+      EnumDecl_create(
+        self.ctx,
+        self.declContext,
+        self.bridgedSourceLoc(for: node.enumKeyword),
+        name,
+        nameLoc,
+        node.genericParameterClause.map(self.visit)?.rawValue,
+        inheritedTypes,
+        node.genericWhereClause.map(self.visit)?.rawValue
+      )
+    }
```
I think it's a great idea to use BumpPtrAllocator.

> we should maybe not only consider an optimization for the case that `count == 1`

I wish we can still have a optimization for `count == 1` case (i.e. not allocating any heap memory,) but I don't think we have a "safe" way to do it. Allocating one element sounds like a waste, but it's a reasonable trade-off.

@ahoppen what was your idea to get a bridged array from `OftenSmallArray<Element>.one(element)` case without using `withXXX()` pattern?

> We could even consider resetting the bump allocator at certain points.

I don't think we need this. And the lifetime of `ASTGenVisitor` is only for creating a AST for each SourceFile, and the memory allocation would always be less than the result AST. So I don't think the allocator would waste much memory for unreasonably long time.

ahoppen replied:
> @ahoppen what was your idea to get a bridged array from `OftenSmallArray<Element>.one(element)` case without using `withXXX()` pattern?

I don’t have an idea, probably should have mentioned that. But I thought that we could have a function that gets bridged arrays for 1, 2, 3, …  `OftenSmallArray`s so that this would read something like the following, which I think would be an improvement. With `OftenSmallArray`, this would read as follows:

```swift
let inheritedTypes = node.inheritanceClause?.inheritedTypeCollection.map(in: self) {
  self.visit($0.typeName).rawValue
} ?? []
return withBridgedArray(inheritedTypes, /* could also take 2 arguments */) { inheritedTypes in
  return EnumDecl_create(…)
}
```

> I don't think we need this. The lifetime of `ASTGenVisitor` is only for creating a AST for each SourceFile, and the memory allocation would always be less than the result AST. So I don't think the allocator would waste much memory for unreasonably long time.

The main point I wanted to make here was that *if* we discover that the bump allocator uses measurable memory we *could* make such an optimization, which should significantly reduce the memory overhead of it even further. 

I also realized that I didn’t mention that we do have a `BumpPtrAllocator` implementation in swift-syntax that can be used here: https://github.com/apple/swift-syntax/blob/main/Sources/SwiftSyntax/BumpPtrAllocator.swift

rintaro replied:
>  I thought that we could have a function that gets bridged arrays for 1, 2, 3, …  OftenSmallArrays

I see, that way we can avoid nested `withBridgedArray` call, yes that's an improvement. But I don't think it's worth it. Allocating one pointer in `BumpPtrAllocator` is not that heavy, readability is more important. :+1:

AnthonyLatsis replied:
I went with bump allocation as suggested. The sequence transforming methods I had to overload are `map`, `compactMap`, and `reversed`, which I’m frankly not too happy about. I wish we could get away with extending `LazySequenceProtocol` with a `toBridgedArrayRef(in:)` method instead. The problem is we need to at least know the max count for allocation, but there is no such API and getting the `count` on lazy wrappers can be painfully expensive.

ahoppen commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -1,162 +1,503 @@
-    let oldDeclContext = declContext
-    declContext = out.declContext
-    defer { declContext = oldDeclContext }
+    return .decl(
+      TypeAliasDecl_create(
+        self.ctx,
+        self.declContext,
+        self.bridgedSourceLoc(for: node.typealiasKeyword),
+        name,
+        nameLoc,
+        node.genericParameterClause.map(self.visit)?.rawValue,
+        self.bridgedSourceLoc(for: node.initializer.equal),
+        self.visit(node.initializer.value).rawValue,
+        node.genericWhereClause.map(self.visit)?.rawValue
+      )
+    )
```
Ah, I see. I’m fine with doing it in a follow-up PR as well. I just want to be clear what the direction that we want to go in, is.

AnthonyLatsis commented on include/swift/AST/CASTBridging.h:
```diff
@@ -302,9 +302,14 @@ void *ArrayExpr_create(BridgedASTContext cContext, BridgedSourceLoc cLLoc,
-void *VarDecl_create(BridgedASTContext cContext, void *nameExpr, void *initExpr,
-                     BridgedSourceLoc cLoc, _Bool isStatic, _Bool isLet,
-                     BridgedDeclContext cDeclContext);
+void *VarDecl_create(
```
> I'd prefer this formatted with clang-format

It was — my config file is stricter that the one that comes with the repo. For context, I’ve grown to dislike this format for several reasons:
* Makes some diffs really hard to read for me (on GitHub, where I cannot customize them)
* Makes history inspection harder, e.g. `git blame`
* Results in large amounts of conflicts when reverting stuff as an alternative to bisecting

Though I guess I should bite the bullet and just be consistent for the sake of the next person to edit this.

AnthonyLatsis commented on lib/AST/CASTBridging.cpp:
```diff
@@ -300,22 +278,42 @@ void *BooleanLiteralExpr_create(BridgedASTContext cContext, bool value,
-      sourceLoc, pattern, sourceLoc, (Expr *)initExpr, declContext);
+      convertSourceLoc(cBindingKeywordLoc),
+      pattern,
+      /*EqualLoc=*/SourceLoc(),
```
It is indeed a problem, but tending to storage decl translation was beyond my intentions for this PR. Point noted though, I will add `FIXME`s for these.

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -1,162 +1,503 @@
+        nameLoc,
+        node.genericParameterClause.map(self.visit)?.rawValue,
+        self.bridgedSourceLoc(for: node.initializer.equal),
+        self.visit(node.initializer.value).rawValue,
+        node.genericWhereClause.map(self.visit)?.rawValue
```
Sounds good. Would you mind if I make this overload generic (and always-inline) to avoid duplication?

ahoppen replied:
If you have an idea of how to make that work generically, I definitely don’t mind. I thought it wouldn’t work because what would you constrain the generic type by? But I’m open for ideas.

AnthonyLatsis replied:
Never mind, I thought of constraining it to `SyntaxProtocol`, but that implies a wasteful call to the dispatch method. Besides, the result need not be an `ASTNode`.

AnthonyLatsis commented on lib/AST/CASTBridging.cpp:
```diff
@@ -365,51 +374,140 @@ void *BraceStmt_create(BridgedASTContext cContext, BridgedSourceLoc cLBLoc,
+BridgedDeclContextAndDecl DestructorDecl_create(
+    BridgedASTContext cContext,
+    BridgedDeclContext cDeclContext,
+    BridgedSourceLoc cDeinitKeywordLoc) {
```
Ben persuaded me to reformat these using the config file that comes with the repo, so this is no longer relevant. For context, I originally formatted these using a custom config file of mine because [comment](https://github.com/apple/swift/pull/67111#discussion_r1264281821). Line breaks before `)` are off by default — you can turn them on.

bnbarham replied:
Ah sorry for not replying to that comment. I don't disagree with your points there, mostly I would just prefer consistency unless we have wider discussions on formatting. Going with what LLVM uses is just the simplest default right now 😅

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -1,162 +1,503 @@
     let isStatic = false  // TODO: compute this
     let isLet = node.bindingKeyword.tokenKind == .keyword(.let)
 
     // TODO: don't drop "initializer" on the floor.
```
[Apparently not](https://github.com/apple/swift/commit/f85d6872782b#diff-e513d3adfa93904a74cb6d3e8ac3ae4701ed41ccaa0461428708231d8aee5b63R78), will remove.

AnthonyLatsis committed:
[NFC] ASTGen: Add Swift names to bridging functions for decls

AnthonyLatsis commented on include/swift/AST/CASTBridging.h:
```diff
@@ -319,20 +326,49 @@ void *BraceStmt_create(BridgedASTContext cContext, BridgedSourceLoc cLBLoc,
-                BridgedArrayRef params, BridgedSourceLoc cParamRLoc,
-                void *_Nullable returnType, BridgedDeclContext cDeclContext);
-void FuncDecl_setBody(void *fn, void *body);
+void *ParamDecl_create(
+    BridgedASTContext cContext,
+    BridgedDeclContext cDeclContext,
+    BridgedSourceLoc cSpecifierLoc,
+    BridgedIdentifier cFirstName,
+    BridgedSourceLoc cFirstNameLoc,
+    BridgedIdentifier cSecondName,
+    BridgedSourceLoc cSecondNameLoc,
+    void *_Nullable opaqueType,
+    void *_Nullable opaqueDefaultArgumentExpr);
```
That‘s an awesome feature, TIL.

ahoppen replied:
What do you mean by the *stringizing*?

AnthonyLatsis replied:
Stringizing is what prepending a macro argument with `#` is called. Suppose `SWIFT_NAME(foo(a:b:))` is long enough to exceed the 80-column limit and `clang-format` adds a line break somewhere in the name, e.g.

```
SWIFT_NAME(foo(a:
:b))
```

This expansion will replace the line break with a space in the resulting string constant:

```
__attribute__((swift_name("foo(a: b:)"))
```

ahoppen replied:
Oh, I see. I just copied the macro from somewhere but we might need to adjust it, I agree.

AnthonyLatsis replied:
Attributed just the new and `*Decl_create` functions for now per the intended scope of this PR. Happy to incrementally audit the rest separately.

ahoppen replied:
Thanks. They look a lot cleaner already!

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -1,162 +1,503 @@
+    )
+  }
+
+  private func visitMemberDeclBlock(
+    _ memberBlock: MemberDeclBlockSyntax,
+    declContextAndDecl: BridgedDeclContextAndDecl
+  ) {
```
Actually I feel like replacing these private methods with  a `withDeclContext` method that takes a closure. Does that sound good to you?

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -1,162 +1,503 @@
+    let (name, nameLoc) = self.bridgedIdentifierAndSourceLoc(for: node.identifier)
+
+    let declAndDeclContext = (node.primaryAssociatedTypeClause?.primaryAssociatedTypeList).map {
+      let (name, nameLoc) = self.bridgedIdentifierAndSourceLoc(for: $0.name)
+      return BridgedIdentifierAndSourceLoc(name: name, nameLoc: nameLoc)
```
I’m going to define this as an `bridgedIdentifierAndSourceLoc` overload on `TokenSyntax` for consistency.

AnthonyLatsis committed:
[NFC] ASTGen: Define `visit(_:)` overloads that take an optional to achieve consistent `self.visit` recursion pattern

AnthonyLatsis committed:
[NFC] ASTGen: Move `visit` methods from "Misc.swift" to "ASTGen.swift"

Also, delete `func visit(_ node: LabeledExprSyntax)` as unused. We directly
visit its children instead, since it’s more than an expression.

AnthonyLatsis committed:
[NFC] ASTGen: Rename "Misc.swift" → "Bridge.swift"

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/Misc.swift:
```diff
@@ -76,3 +61,120 @@ extension String {
+    }
+  }
+
+  /// Given a token, obtains a uniqued `ASTContext`-owned copy of its text, and its start source location in the
+  /// current file, skipping leading trivia.
+  @inline(__always)
+  func bridgedIdentifierAndSourceLoc(for token: TokenSyntax?) -> (BridgedIdentifier, BridgedSourceLoc) {
+    return (self.bridgedIdentifier(for: token), self.bridgedSourceLoc(for: token))
+  }
```
I gave this quite a bit of thought and I didn’t wind up with a strong opinion, but perhaps you’re right about the calls being more idiomatic on `TokenSyntax`; it was tempting to make it a method on `ASTGenVisitor` because it only ever takes `self`.

AnthonyLatsis committed:
ASTGen: Start diagnosing instances of an invalid syntax tree

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -1,162 +1,503 @@
+        default: fatalError("Unknown precedence group relation attribute")
+        }
+      case .precedenceGroupAssignment(let node):
+        partialResult.assignment = node
```
I chose to define a separate error for diagnosing duplicate syntax.

ahoppen commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -1,162 +1,503 @@
-    let secondName: BridgedIdentifier
+extension ASTGenVisitor {
+  public func visit(_ node: FunctionDeclSyntax) -> ASTNode {
+    // FIXME: Compute this location
+    let staticLoc: BridgedSourceLoc = nil
```
Sure

AnthonyLatsis commented:
@swift-ci please smoke test macOS

ahoppen reviewed (commented):
Thank you so much @AnthonyLatsis. ASTGen looks sooooo much better with these changes.  😍 

I left a few inline comments with the expectation that all of them are minor. I will file issues for any bigger changes that I think could be made to improve ASTGen further.

---

I mostly reviewed the Swift changes in lib/ASTGen. @hamishknight could you take a look at the C++ side? In particular the changes in lib/AST/CASTBridging.cpp and lib/AST/Decl.cpp but really I’d appreciate your review on anything outside of lib/ASTGen.

ahoppen commented:
I added a new issue label for the improvements that I think we should make to ASTGen: https://github.com/apple/swift/labels/ASTGen

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/Generics.swift:
```diff
@@ -38,62 +34,45 @@ extension ASTGenVisitor {
+          SecondType: self.visit(sameType.rightType).rawValue
+        )
+      case .layoutRequirement(_):
+        fatalError("Cannot handle layout requirements!")
```
Sure.

AnthonyLatsis commented on include/swift/AST/CASTBridging.h:
```diff
@@ -331,42 +373,186 @@ void *UnresolvedDotExpr_create(BridgedASTContext cContext, void *base,
+    BridgedArrayRef cInheritedTypes, void *_Nullable opaqueGenericWhereClause,
+    BridgedSourceRange cBraceRange);
+
+typedef enum ENUM_EXTENSIBILITY_ATTR(open) {
```
I simply followed the existing pattern really. `closed` exports the enum as `@frozen`. `open` *probably* exists as a requisite for importing C enums that are not `closed` into a Swift module compiled with library evolution enabled, to avoid accidental openness. If binary compatibility is not a concern in the foreseeable future, we can indeed use `closed` or just remove the attribute altogether.

AnthonyLatsis replied:
Ok [I looked this up](https://github.com/apple/swift/blob/f5407533ee8477f5bee381a0c4a95ac6ac92b410/docs/HowSwiftImportsCAPIs.md?plain=1#L946). We need the attribute or else the enum will be imported as a struct. I will mark them as closed then. As for `BridgedImportKind`, I suggest keeping it in sync with `ImportKindSpecifierOptions` (open).

ahoppen replied:
Thanks 👍🏽

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/ASTGen.swift:
```diff
@@ -112,18 +123,206 @@ struct ASTGenVisitor: SyntaxTransformVisitor {
+    visit(Syntax(node.decl))
+  }
+
+  public func visit(_ node: LabeledExprSyntax) -> ASTNode {
+    visit(node.expression)
+  }
```
Yeah, no direct calls to this anywhere, we visit the children of a `LabeledExprSyntax` instead. Should I just delete it with a comment in that commit?

ahoppen replied:
Yes, I think we should delete `func visit(_ node: LabeledExprSyntax)` for now.

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/ASTGen.swift:
```diff
@@ -112,18 +123,206 @@ struct ASTGenVisitor: SyntaxTransformVisitor {
+      return .init()
+    }
+
+    let baseAddress = astgen.allocator.allocate(T.self, count: self.count).baseAddress!
+    do {
+      var currentAddress = baseAddress
+      for element in self {
+        currentAddress.initialize(to: transform(element))
+        currentAddress += 1
+      }
+    }
```
I trust you mean `self.lazy.map(transform)`? What made me reluctant to pull on this low-hanging fruit is how it bloats the optimized SIL. We aren’t doing a great job at optimizing away the lazy wrapper. I don’t mind shortening it if you believe the tradeoff is acceptable.

ahoppen replied:
Oh, I was assuming that it compiled to about the same code. If your implementation is faster than the shorter one, definitely keep it. I would just add a comment which explains that we don’t use `initialize(from:)` for performance reasons.

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/Diagnostics.swift:
```diff
@@ -458,3 +456,111 @@ public func renterQueuedDiagnostics(
 }
+
+
+// MARK: - ASTGen diagnostics
```
`ASTGenError.swift` or `ASTGenDiagnostics.swift`?

AnthonyLatsis replied:
IMO `ASTGenDiagnostics.swift` is self-explanatory and more prudent . I also propose that we rename  `Diagnostics.swift` to either `DiagnosticsBridging.swift` or `DiagnosticsBridge.swift` because it mostly, if not exclusively, deals with emitting SwiftSyntax diagnostics via the C++ engine.

ahoppen replied:
Let’s do `ASTGenDiagnostics.swift`. That way we don’t need to rename the file if we introduce ASTGen warnings at some point.



> I also propose that we rename `Diagnostics.swift` to either `DiagnosticsBridging.swift`.

Very good idea 👍🏽

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/ASTGen.swift:
```diff
@@ -112,18 +123,206 @@ struct ASTGenVisitor: SyntaxTransformVisitor {
+
+  @inline(__always)
+  func visit(_ node: GenericWhereClauseSyntax?) -> ASTNode? {
+    node.map(self.visit)
```
Ooh, totally worth it.

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -1,165 +1,528 @@
-      self.ctx, self.declContext, aliasLoc, equalLoc, name, nameLoc, genericParams)
+    let (name, nameLoc) = node.name.bridgedIdentifierAndSourceLoc(in: self)
 
-    let oldDeclContext = declContext
-    declContext = out.declContext
-    defer { declContext = oldDeclContext }
```
Apparently this is only needed for code completion: #16394. Nodes that need a decl context cannot appear in the RHS of a type alias.

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/Diagnostics.swift:
```diff
@@ -458,3 +456,111 @@ public func renterQueuedDiagnostics(
+
+// MARK: - Invalid syntax tree diagnostics
+
+// FIXME: Make 'SwiftParser.TokenSpecSet' public?
+typealias TokenSpecSet = CaseIterable & CustomStringConvertible
+
+/// An error emitted when a token does not match any of several expected kinds.
+struct ExpectedTokenKindToBeOneOfError<Kinds: TokenSpecSet>: ASTGenError {
+  private let baseError: UnexpectedTokenKindError
+
+  var token: TokenSyntax {
+    baseError.token
+  }
+
+  init(token: TokenSyntax) {
+    self.baseError = .init(token: token)
+  }
+
+  var message: String {
+    let kinds = Kinds.allCases.lazy.map { "'\($0.description)'" }.joined(separator: ", ")
+    return """
+      \(self.baseError.message)
+      expected one of: \(kinds)
+      """
+  }
+}
```
No, I added it in anticipation of a public `TokenSpecSet`. Would you rather have it added later? Alternatively, I could use it against the bridged enums as a temporary measure.

ahoppen replied:
Let’s add it when we actually use start using it.

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -1,165 +1,528 @@
+        let keyword = node.higherThanOrLowerThanLabel
+        switch keyword.tokenKind {
+        case .keyword(.higherThan):
+          guard let current = partialResult.higherThanRelation else {
+            partialResult.higherThanRelation = node
+            return
+          }
```
I’d be grateful to keep it — I like `reduce` in that the result can be a constant. Should read better once I rename `partialResult` to `body`.

ahoppen replied:
OK, let’s keep `reduce`. I don’t have super strong feelings about it and it’s probably that I just didn’t really use `reduce` in this way before.

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -1,165 +1,528 @@
-    if let body = body {
-      FuncDecl_setBody(out.funcDecl, body.rawValue)
+    let associativityValue: BridgedAssociativity
+    switch body.associativity?.value {
+    case let token?:
+      guard let value = BridgedAssociativity(from: token.tokenKind) else {
+        self.diagnose(Diagnostic(node: token, message: UnexpectedTokenKindError(token: token)))
+        fallthrough
+      }
+
+      associativityValue = value
+    case nil:
+      associativityValue = .none
     }
```
Sounds good. This is one of those trade-offs where lightweight control flow looks better but for a pesky duplication or whatnot. I wish we had an idiomatic control flow convenience for this pattern.

AnthonyLatsis committed:
[NFC] ASTGen: Rename "Diagnostics.swift" → "DiagnosticsBridge.swift"

This file deals with emitting SwiftSyntax diagnostics via the C++ engine and
rendering C++ diagnostics via SwiftSyntax rather than ASTGen diagnostics.

hamishknight commented on lib/AST/CASTBridging.cpp:
```diff
@@ -378,51 +382,109 @@ void *BraceStmt_create(BridgedASTContext cContext, BridgedSourceLoc cLBLoc,
+    cSecondNameLoc = cFirstNameLoc;
+  }
+
+  auto *defaultValue = static_cast<Expr *>(opaqueDefaultValue);
+  const auto defaultArgumentKind = getDefaultArgKind(defaultValue);
+  if (defaultArgumentKind == DefaultArgumentKind::Inherited) {
+    defaultValue = nullptr;
+  }
```
Presumably this is handling the `= super` default argument case for swift interfaces? I think we'll only want to apply this for swift interface source files to match what the old parser does. For regular source files, this ought to remain `DefaultArgumentKind::Normal`, and we should diagnose the bare use of `super` to match the old parser.

AnthonyLatsis replied:
Yeah, I messed it up, thanks for spotting! Let’s just leave out the handling of this case for now. I think the right thing to do is compute `DefaultArgumentKind` and diagnose `super` during Sema. `ASTGen` shouldn’t to go out of its way to do this.

AnthonyLatsis replied:
#70579.

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/ASTGen.swift:
```diff
@@ -112,18 +123,206 @@ struct ASTGenVisitor: SyntaxTransformVisitor {
+  }
+}
+
+extension Optional where Wrapped: SyntaxCollection {
+  /// Like ``Sequence.map(_:)``, but returns a `BridgedArrayRef` with a lifetime tied to that of `astgen`.
+  @inline(__always)
+  func map<T>(in astgen: ASTGenVisitor, _ transform: (Wrapped.Element) -> T) -> BridgedArrayRef {
+    switch self {
+    case .none: return .init()
+    case .some(let unwrapped): return unwrapped.map(in: astgen, transform)
+    }
+  }
+}
```
> I’m not a huge fan of this method because there’s no corresponding stdlib equivalent

What about the methods under `extension Optional<TokenSyntax>`? They’re essentially the same convenience. I think what sets these methods apart is that their result types are arguably `ExpressibleByNilLiteral` (`BridgedArrayRef` is currently not but could follow the precedent set by `BridgedSourceLoc` and `BridgedIdentifier`). Supposing they are, `?? nil` ideally calls for democratization of optional chaining such that it works (flattening and all) with custom `ExpressibleByNilLiteral` types. Another way to make this work for our purposes is to define a custom `map`/`flatMap` on `Optional`, hence the method.

ahoppen replied:
I stumbled a bit as well when I read `extension `Optional<TokenSyntax>` but then accepted them because `BridgedIdentifier` is `ExpressibleByNilLiteral` and `?? nil` looks weird. 

I see your argument that you could view `BridgedArrayRef` as `ExpressibleByNilLiteral` now. I’m fine with keeping it.

AnthonyLatsis commented on lib/ASTGen/Sources/ASTGen/ASTGen.swift:
```diff
@@ -112,18 +123,206 @@ struct ASTGenVisitor: SyntaxTransformVisitor {
+  }
+}
+
+extension Collection {
+  /// Like ``Sequence.reversed()``, but returns a `BridgedArrayRef` with a lifetime tied to that of `astgen`.
+  func reversed(in astgen: ASTGenVisitor) -> BridgedArrayRef {
+    if self.isEmpty {
+      return .init()
+    }
+
+    let baseAddress = astgen.allocator.allocate(Element.self, count: self.count).baseAddress!
+
+    do {
+      var currentAddress = baseAddress
+      for element in self.lazy.reversed() {
+        currentAddress.initialize(to: element)
+        currentAddress += 1
+      }
+    }
+
+    return .init(data: baseAddress, numElements: SwiftInt(self.count))
+  }
+}
+
```
> ```diff
> - let memberComponents = reverseMemberComponents.reversed(in: self)
> + let memberComponents = reverseMemberComponents.reversed().map(in: self) { $0 }
> ```

I really want to avoid intermediate arrays, mostly because of the performance smell but in a way I also see them as defeating the purpose of bump allocation—we would still be reference-counting and allocating just as much.

> As a bonus, we could define 
> 
> ```swift
> extension Collection {
>   func bridgedArray(in astgen: ASTGenVisitor) -> BridgedArrayRef {
>     return self.map(in: astgen) { $0 }
>   } 
> }
> ```

I [had a very similar idea](https://github.com/apple/swift/pull/67111#discussion_r1316084733) that enforces lazy array bridging. This helps to prevent eager operations to a certain extent by guiding the programmer. The problem of doing this on `LazyCollectionProtocol` is we need a sensible upper bound for how much to allocate, but using `count` on lazy wrappers can be very costly for compact maps and such. Would you be up for this sort of hack:

1. Add a `ASTGenLazyCollection` protocol with a single `maxCount` requirement.
1. The default implementation of `maxCount` can just return `count`.
1. Conform the standard library lazy wrappers we use to `ASTGenLazyCollectionProtocol`
1. ```swift
   extension ASTGenLazyCollectionProtocol {
     func bridgedArray(in astgen: ASTGenVisitor) -> BridgedArrayRef {
       if self.isEmpty {
         return .init()
       }
    
       let buffer = astgen.allocator.allocate(Element.self, count: self.maxCount)
       _ = buffer.initialize(from: self)
    
       return .init(data: buffer.baseAddress!, numElements: SwiftInt(self.maxCount))
     }
   }
   ```

This trades the performance boost from defining custom operations vs. using lazy wrappers for a consistent, low-maintenance, and more or less future-proof pattern.

ahoppen replied:
> I really want to avoid intermediate arrays, mostly because of the performance smell but in a way I also see them as defeating the purpose of bump allocation—we would still be reference-counting and allocating just as much.

Ah, damn. I forgot about the intermediate array creation while reviewing this. I’m fine with keeping it. Could you add a doc comment to `reversed(in:)` that highlights that the method exists for exactly this performance reason?

> Conform the standard library lazy wrappers we use to `ASTGenLazyCollectionProtocol`

Which types would you conform to `ASTGenLazyCollectionProtocol`? For example `array.lazy.reversed()` is a `ReversedCollection<LazySequence<UnderlyingSequence>>` and I don’t think we can guarantee efficient `maxCount` access on that.

AnthonyLatsis replied:
> Which types would you conform to `ASTGenLazyCollectionProtocol`? For example `array.lazy.reversed()` is a `ReversedCollection<LazySequence<UnderlyingSequence>>` and I don’t think we can guarantee efficient `maxCount` access on that.

Arrgh! `ReversedCollection._base` is public, which made me think the bases in other wrappers are too. I guess this nut is unlikely to ever crack nicely unless we integrate `maxCount` into the stdlib like so:

* `LazySequenceProtocol` gets a `base` requirement
* `Sequence` gets a `maxCount` requirement with default implementations of  `Int.max` on `Sequence`, `self.count` on `Collection`, and `base.maxCount` on `LazySequenceProtocol`.
* `ReversedCollection` conforms to `LazySequenceProtocol`

Since (as Rintaro pointed out below) `reversed` is already lazy on `BidirectionalCollection`, let’s do `Collection.bridgedArray(in:)` and make an exception for `compactMap`. As much as I dislike having to keep `compactMap(in:)`, I would rather not have us calling `count` on a lazy `compactMap`.

ahoppen replied:
Sounds good 👍🏽. I can’t tell you why but `compactMap` also triggered me a lot less than `reversed`.

AnthonyLatsis commented:
> https://github.com/apple/swift/labels/ASTGen

Much appreciated!

hamishknight commented on lib/ASTGen/Sources/ASTGen/Decls.swift:
```diff
@@ -75,43 +99,42 @@ extension ASTGenVisitor {
-    let funcLoc = bridgedSourceLoc(for: node.funcKeyword)
-    let (name, nameLoc) = node.name.bridgedIdentifierAndSourceLoc(in: self)
+    // FIXME: Compute this location
+    let staticLoc: BridgedSourceLoc = nil
```
Gah we really need to turn the staticness into a decl attribute, I started working on that a little while back but never got the time to finish it

ahoppen reviewed (approved):
Sorry for the long reviews on this PR, this looks good to me now. The only (very minor) comment that’s still open is https://github.com/apple/swift/pull/67111#discussion_r1317500453 but I’m happy to postpone that to a follow-up PR. If you want to do that, could you file an issue for it?

I would still like to get a review from @hamishknight on the C++ AST side of things.

hamishknight reviewed (approved):
Great stuff!

rintaro commented on lib/ASTGen/Sources/ASTGen/ASTGen.swift:
```diff
@@ -112,18 +123,224 @@ struct ASTGenVisitor: SyntaxTransformVisitor {
+    }
+
+    let buffer = astgen.allocator.allocate(Element.self, count: self.count)
+    _ = buffer.initialize(from: self.lazy.reversed())
```
I've just read your previous discussion with Alex. But `.reversed` is guaranteed _not_ to create new storage.
https://developer.apple.com/documentation/swift/reversedcollection
> * `c.reversed()` does not create new storage

So I don't think we need this method. Please just do
```swift
extension Collection {
  func bridgedArray(in astgen: ASTGenVisitor) -> BridgedArrayRef {
    let buffer = astgen.allocator.allocate(Element.self, count: self.count)
    _ = buffer.initialize(from: self)
    return .init(data: buffer.baseAddress, numElements: SwiftInt(self.count))
  } 
}
```
or
```swift
extension BridgedArrayRef {
  init<C: Collection>(_ collection: C, in astgen: ASTGenVIsitor) {
    let buffer = astgen.allocator.allocate(C.Element.self, count: collection.count)
    _ = buffer.initialize(from: collection)
    self.data = buffer.baseAddress
    self.numElements = SwiftInt(buffer.count)
  }
}
```
And use `.reversed()` at the cal site.

AnthonyLatsis replied:
> `.reversed()` on `BidirectionalCollection` directly doesn't create an intermediate array

Thank you, I forgot about this!

AnthonyLatsis replied:
@rintaro I would like to statically enforce calling `bridgedArray` on lazy collections to at least a certain extent so that we don’t have to keep our eyes as open for these intermediate arrays. Do you mind having this instead (with  `map(in:)` removed)?:

```swift
extension LazyCollectionProtocol {
  func bridgedArray(in astgen: ASTGenVisitor) -> BridgedArrayRef {
    ...
  }
}
// 'ReversedCollection' does not conform to 'LazyCollectionProtocol', and cannot here because it only
// conditionally conforms to 'LazySequenceProtocol' in the standard library.
// FIXME: We could make it conform unconditionally
extension ReversedCollection {
  func bridgedArray(in astgen: ASTGenVisitor) -> BridgedArrayRef {
    self.lazy.bridgedArray(in: astgen)
  }
}
```

AnthonyLatsis commented:
@swift-ci please smoke test

AnthonyLatsis commented:
@ahoppen We have a problem with `swift_name` on Windows. I thought we were using the just-built clang to compile the Swift compiler, but apparently not.

AnthonyLatsis commented:
@swift-ci please test Windows

AnthonyLatsis commented:
I take it the conditional def works because we skip the early swift-syntax on Windows. Is this an OK workaround?

bnbarham commented:
> I take it the conditional def works because we skip the early swift-syntax on Windows. Is this an OK workaround?

That's not going to be true very soon, but regardless the conditional def seems fine to me - it isn't needed when compiling C++, only when importing into Swift (in which case, it will be defined). Unless I'm misunderstanding something crucial here?

ahoppen commented:
@swift-ci Please test

AnthonyLatsis commented:
> it isn't needed when compiling C++, only when importing into Swift (in which case, it will be defined). Unless I'm misunderstanding something crucial here?

Nah, actually, thank you, I had a brain fart. I just realized the importer invokes clang to parse headers (so of course it will be defined when importing into Swift). For some reason I had this silly impression that it was C compiler-agnostic even though I knew the importer reads clang ASTs 🙃

AnthonyLatsis commented:
@ahoppen @DougGregor @bnbarham @hamishknight Thank you so much for weighing in!

bnbarham commented:
Woo, it's in! Thanks for all the work here @AnthonyLatsis!

ahoppen commented:
Thank **you** for doing all this work! 🙏🏽
